samplers/aws/xray/internal/manifest.go

```diff
@@ -0,0 +1,353 @@
+			continue
+		}
+
+		// create rule and store it in temporary manifest to avoid thread safety issues
+		tempManifest.createRule(*records.SamplingRule)
+	}
+
+	// Re-sort to fix matching priorities.
+	tempManifest.sort()
+
+	m.mu.Lock()
+	m.Rules = tempManifest.Rules
+	m.refreshedAt = m.clock.Now()
+	m.mu.Unlock()
+}
+
+func (m *Manifest) createRule(ruleProp ruleProperties) {
+	cr := reservoir{
+		capacity: ruleProp.ReservoirSize,
+		mu:       &sync.RWMutex{},
+	}
+
+	csr := Rule{
+		reservoir:          cr,
+		ruleProperties:     ruleProp,
+		samplingStatistics: &samplingStatistics{},
+	}
+
+	m.Rules = append(m.Rules, csr)
+}
+
+func (m *Manifest) updateTargets(targets *getSamplingTargetsOutput) (refresh bool, err error) {
+	// update sampling targets for each rule
+	for _, t := range targets.SamplingTargetDocuments {
+		if err := m.updateReservoir(t); err != nil {
+			return false, err
+		}
+	}
+
+	// consume unprocessed statistics messages
+	for _, s := range targets.UnprocessedStatistics {
+		m.logger.V(5).Info(
+			"error occurred updating sampling target for rule, code and message", "RuleName", s.RuleName, "ErrorCode",
+			s.ErrorCode,
+			"Message", s.Message,
+		)
+
+		// do not set any flags if error is unknown
+		if s.ErrorCode == nil || s.RuleName == nil {
+			continue
+		}
+
+		// set batch failure if any sampling statistics returned 5xx
+		if strings.HasPrefix(*s.ErrorCode, "5") {
+			return false, fmt.Errorf("sampling statistics returned 5xx")
+		}
+
+		// set refresh flag if any sampling statistics returned 4xx
+		if strings.HasPrefix(*s.ErrorCode, "4") {
+			refresh = true
+		}
+	}
+
+	// set refresh flag if modifiedAt timestamp from remote is greater than ours
+	if remote := targets.LastRuleModification; remote != nil {
+		// convert unix timestamp to time.Time
+		lastRuleModification := time.Unix(int64(*targets.LastRuleModification), 0)
+
+		if lastRuleModification.After(m.refreshedAt) {
+			refresh = true
+		}
+	}
+
+	return
+}
+
+func (m *Manifest) updateReservoir(t *samplingTargetDocument) (err error) {
+	if t.RuleName == nil {
+		return fmt.Errorf("invalid sampling target. Missing rule name")
+	}
+
+	if t.FixedRate == nil {
+		return fmt.Errorf("invalid sampling target for rule %s. Missing fixed rate", *t.RuleName)
+	}
+
+	for index := range m.Rules {
+		if m.Rules[index].ruleProperties.RuleName == *t.RuleName {
+			m.Rules[index].reservoir.refreshedAt = m.clock.Now()
+
+			// Update non-optional attributes from response
+			m.Rules[index].ruleProperties.FixedRate = *t.FixedRate
+
+			// Update optional attributes from response
+			if t.ReservoirQuota != nil {
+				m.Rules[index].reservoir.quota = *t.ReservoirQuota
+			}
+			if t.ReservoirQuotaTTL != nil {
+				m.Rules[index].reservoir.expiresAt = time.Unix(int64(*t.ReservoirQuotaTTL), 0)
+			}
+			if t.Interval != nil {
+				m.Rules[index].reservoir.interval = time.Duration(*t.Interval)
+			}
+		}
```

Comment:
You need to take the Rule's lock to update these fields, or Rule needs an update method that captures that lock.

Reply:
Since we're doing deep copy [here ](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1859/files/0cf3224c740202e79bd33649f307941f9954ee55#diff-d909e7bd591a0ccb2cedd3b1602921ca5ca1612d282189172fb3d666fffdcfd0R127) we don't need lock here while updating reservoir values. While [updating the ruleProperties](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1859/files/0cf3224c740202e79bd33649f307941f9954ee55#diff-d909e7bd591a0ccb2cedd3b1602921ca5ca1612d282189172fb3d666fffdcfd0R174) we are creating manifest every time so don't think we need lock there either.